### ISSUE TITLE
Python interface: allow the retrieval of a variable using `<component>/<variable>[/prime...]`

### ISSUE THREAD

**Issue by jonrkarr** [opencor/opencor#2612]
## Motivation
This question is motivated by some of the SED-ML files in Physiome such as
- `e/4f1`: [`1998_nash.sedml`](https://models.physiomeproject.org/workspace/4f0/file/746a9569a42515d549497d72cb6abd74ea63dfba/1998_nash.sedml)
- `e/4ea`
- `e/430`

## Background
We are executing SED-ML+CellML files in the following steps:
1. Interpreting the SED-ML
2. Breaking down the SED-ML into individual tasks
3. Creating a SED-ML file for each individual task
4. Using OpenCOR/Python to execute each task
5. Collecting the results of each task (value of each target of each variable of each data generator)
6. Assembling the results into the requested outputs (reports and plots)

We don't use OpenCOR to execute entire SED-ML files because this strategy supports almost all of SED-ML L1V3, rather than the subset that OpenCOR supports. This strategy works for all SED-ML files in Physiome except for those outlined above.

However, this strategy doesn't work for a few SED-ML files where some of the targets of variables of data generators are not exported by OpenCOR. For example, [`1998_nash.sedml`](https://models.physiomeproject.org/workspace/4f0/file/746a9569a42515d549497d72cb6abd74ea63dfba/1998_nash.sedml) needs the value of the target `/cellml:model/cellml:component[@name='active_contraction']/cellml:variable[@name='T']`. However, OpenCOR `sim.results()` doesn't include the value of this variable because its part of a component (`active_contraction`) which is encapsulated into a group.

## Question
How can the values of such targets be retrieved? 
- Is there another method to retrieve their value?
- Is the SED-ML encoding or CellML encapsulation incorrect?

**Comment by agarny** [comment]
I just had a look at [`1998_nash.sedml`](https://models.physiomeproject.org/workspace/4f0/file/746a9569a42515d549497d72cb6abd74ea63dfba/1998_nash.sedml) and it's all working fine for me in OpenCOR:
![Screen Shot 2022-03-29 at 09 46 38](https://user-images.githubusercontent.com/602265/160484251-1461a77d-2ae5-4e98-8823-58f307a24d55.png)
And I would expect the following Python code to work for you?
```python
import opencor as oc
s = oc.open_simulation('https://models.physiomeproject.org/workspace/4f0/rawfile/746a9569a42515d549497d72cb6abd74ea63dfba/1998_nash.sedml')
s.run()
r = s.results()
v = r.data_store().voi_and_variables()
t = v['interface/active_contraction/T'].values()
```

From there, if you try to check the contents of `t`, you should get:
```python
array([0.00000000e+00, 1.00645170e-10, 6.41030627e-09, 7.29521599e-08,
       4.09764832e-07, 1.56290238e-06, 4.66643419e-06, 1.17664444e-05,
       2.62170302e-05, 5.31482958e-05, 1.00006338e-04, 1.77165212e-04,
       2.98610858e-04, 4.82696958e-04, 7.52972663e-04, 1.13908212e-03,
       1.67773564e-03, 2.41375241e-03, 3.40117430e-03, 4.70445067e-03,
       6.39969342e-03, 8.57600177e-03, 1.13368558e-02, 1.48015779e-02,
       1.91068598e-02, 2.44083549e-02, 3.08823317e-02, 3.87273871e-02,
       4.81662151e-02, 5.94474265e-02, 7.28474153e-02, 8.86722652e-02,
       1.07259688e-01, 1.28980985e-01, 1.54243025e-01, 1.83490223e-01,
       2.17206501e-01, 2.55917231e-01, 3.00191120e-01, 3.50642041e-01,
       4.07930757e-01, 4.72766544e-01, 5.45908657e-01, 6.28167617e-01,
       7.20406287e-01, 8.23540692e-01, 9.38540533e-01, 1.06642937e+00,
       1.20828439e+00, 1.36523576e+00, 1.53846544e+00, 1.72920546e+00,
       1.93873560e+00, 2.16838032e+00, 2.41950506e+00, 2.69351167e+00,
       2.99183303e+00, 3.31592679e+00, 3.66726817e+00, 4.04734181e+00,
       4.45763266e+00, 4.89961587e+00, 5.37474574e+00, 5.88444373e+00,
       6.43008566e+00, 7.01298798e+00, 7.63439351e+00, 8.29545654e+00,
       8.99722755e+00, 9.74063775e+00, 1.05264836e+01, 1.13554115e+01,
       1.22279032e+01, 1.31442618e+01, 1.41045989e+01, 1.51088230e+01,
       1.61566295e+01, 1.72474932e+01, 1.83806618e+01, 1.95551533e+01,
       2.07697544e+01, 2.20230231e+01, 2.33132929e+01, 2.46386808e+01,
       2.59970976e+01, 2.73862609e+01, 2.88037109e+01, 3.02468285e+01,
       3.17128554e+01, 3.31989160e+01, 3.47020405e+01, 3.62191894e+01,
       3.77472780e+01, 3.92832017e+01, 4.08238603e+01, 4.23661823e+01,
       4.39071477e+01, 4.54438100e+01, 4.69733157e+01, 4.84929227e+01,
       5.00000161e+01, 5.14921222e+01, 5.29669200e+01, 5.44222500e+01,
       5.58561218e+01, 5.72667182e+01, 5.86523977e+01, 6.00116957e+01,
       6.13433226e+01, 6.26461612e+01, 6.39192621e+01, 6.51618384e+01,
       6.63732588e+01, 6.75530398e+01, 6.87008379e+01, 6.98164407e+01,
       7.08997576e+01, 7.19508105e+01, 7.29697250e+01, 7.39567202e+01,
       7.49121004e+01, 7.58362457e+01, 7.67296035e+01, 7.75926807e+01,
       7.84260351e+01, 7.92302689e+01, 8.00060212e+01, 8.07539620e+01,
       8.14747858e+01, 8.21692067e+01, 8.28379533e+01, 8.34817636e+01,
       8.41013817e+01, 8.46975537e+01, 8.52710246e+01, 8.58225353e+01,
       8.63528199e+01, 8.68626040e+01, 8.73526021e+01, 8.78235167e+01,
       8.82760362e+01, 8.87108343e+01, 8.91285687e+01, 8.95298804e+01,
       8.99153934e+01, 9.02857139e+01, 9.06414302e+01, 9.09831125e+01,
       9.13113127e+01, 9.16265647e+01, 9.19293842e+01, 9.22202690e+01,
       9.24996993e+01, 9.27681379e+01, 9.30260305e+01, 9.32738059e+01,
       9.35118771e+01, 9.37406406e+01, 9.39604778e+01, 9.41717549e+01,
       9.43748236e+01, 9.45700213e+01, 9.47576720e+01, 9.49380862e+01,
       9.51115619e+01, 9.52783845e+01, 9.54388278e+01, 9.55931541e+01,
       9.57416147e+01, 9.58844502e+01, 9.60218914e+01, 9.61541591e+01,
       9.62814647e+01, 9.64040107e+01, 9.65219911e+01, 9.66355915e+01,
       9.67449896e+01, 9.68503555e+01, 9.69518523e+01, 9.70496358e+01,
       9.71438552e+01, 9.72346536e+01, 9.73221678e+01, 9.74065287e+01,
       9.74878618e+01, 9.75662872e+01, 9.76419198e+01, 9.77148699e+01,
       9.77852427e+01, 9.78531393e+01, 9.79186565e+01, 9.79818868e+01,
       9.80429191e+01, 9.81018383e+01, 9.81587258e+01, 9.82136599e+01,
       9.82667152e+01, 9.83179634e+01, 9.83674733e+01, 9.84153107e+01,
       9.84615387e+01])
```

Note that the `T` variable is declared in the `active_contraction` component which is encapsulated in the `interface` component. So, using OpenCOR's Python interface, you need to provide the full path to `T`, i.e. `interface/active_contraction/T`.

Now, looking at the SED-ML file, I can see that we only provide `active_contraction/T`. I am guessing that this is what is causing you guys problems.

I would say that the SED-ML file is correct and that it's OpenCOR's Python interface which is a bit stricter and requires the full path to a variable. @nickerso, would you agree with that? If so then I will modify OpenCOR's Python interface accordingly (and potentially break some existing Python scripts...?).

**Comment by jonrkarr** [comment]
I saw where variables show up under `interface`. We need a programmatic solution that always works. Because, we're interpreting targets as XML XPaths, the targets need to be valid XPaths (as I believe are intended with SED-ML).

**Comment by agarny** [comment]
Yes, I appreciate that. If we are not to modify our SED-ML files or OpenCOR's Python interface then I guess you could always do something like:
```python
import opencor as oc
s = oc.open_simulation('https://models.physiomeproject.org/workspace/4f0/rawfile/746a9569a42515d549497d72cb6abd74ea63dfba/1998_nash.sedml')
s.run()
r = s.results()
v = r.data_store().voi_and_variables()
t = None
for k in v.keys():
    if 'active_contraction/T' in k:
        t = v[k].values()
        break
if t is not None:
    print(t)
else:
    print('Undefined!')
```

**Comment by jonrkarr** [comment]
We need something that works for any SED-ML file used with CellML+OpenCOR. If this can be done without modifying the SED-ML or OpenCOR, I'd be happy to change our code.

**Comment by agarny** [comment]
I would expect the above to work for any SED-ML file used with CellML + OpenCOR's Python interface.

Still, @nickerso, what's your view on OpenCOR's currenty Python interface? Should the full path to a variable be provided (as is currently the case) or should `<component>/<variable>` be sufficient? If the latter then I should be able to quickly modify the Python interface (but it might break some existing OpenCOR Python scripts).

**Comment by jonrkarr** [comment]
Is `interface/` always the additional prefix that needs to be added?

**Comment by agarny** [comment]
No, OpenCOR's Python interface currently requires you to provide the full path to a variable. In the model, the variable `T` is defined in the component `active_contraction`, BUT... that component is encapsulated in the component `interface`. So, to access the variable `T`, you need to reference `interface/active_contraction/T` when using OpenCOR's Python interface. We did this because this is the way the variable `T` shows up in the OpenCOR GUI. However, and as is the case in the SED-ML file, to reference `active_contraction/T` is (should be!) good enough. This is why I am happy to amend OpenCOR's Python interface so that we can use the short reference to a variable rather than the full one. In fact, maybe OpenCOR's Python interface could support both...? @nickerso

**Comment by nickerso** [comment]
I think the issue is that there is no easy way to map an XPath to an OpenCOR path without parsing the CellML file and working out the required hierarchy. I suspect that the full "path" is required due to the Python interface exposing all variables in a model not just the variables actually defined in the top-level document - and for imported models you need the full path to disambiguate which instance of an imported component is being referenced.

This is not an issue in the SED-ML documents as only variables in the top-level CellML document can have an XPath, but does make mapping to the OpenCOR path painful. Also in the OpenCOR GUI only variables in the top-level model document are presented, so again, no issues with imports.

**Comment by agarny** [comment]
OpenCOR only tracks top-level variables (i.e. what you can see in the GUI). It's just that for the Python interface, we (@dbrnz and/or I) decided to use the full path to a variable and this probably to match what we see in the GUI.

Ok, I guess that for OpenCOR's Python interface, we could support both the full path and `<component>/<variable>` (assuming that it can be done, i.e. a full path is not needed for some reasons that I can't remember). Then, in libOpenCOR, we will only support `<component>/<variable>` (again, if a full path is not needed for some reason or another).

**Comment by agarny** [comment]
@jonrkarr, @nickerso, and @dbrnz: I have just had a look at this issue and, as expected, `v` in https://github.com/opencor/opencor/issues/2612#issuecomment-1081184922 is a Python dictionary that uses the URI of a variable as a key. That URI is the full path to a variable. In other words, for variable `T` which is defined in component `active_contraction` which is encapsulated by component `interface`, the URI is `interface/active_contraction/T`.

The problem with that URI is that, from the SED-ML file, we only know that variable `T` is in component `active_contraction`. So, we don't know that `active_contraction` is encapsulated by `interface`, hence my suggestion in https://github.com/opencor/opencor/issues/2612#issuecomment-1081184922.

Now, my suggestion doesn't account for the fact that SED-ML now supports derivatives (see https://github.com/SED-ML/sed-ml/issues/4). Still, it would be easy to adapt, keeping in mind that if `T` was to be a state variable then the URI for its, say, second derivative would be `interface/active_contraction/T/prime/prime`.

So, different options:
1. **Keep things as they are**, but this means that @jonrkarr and anyone interested in interacting with SED-ML from OpenCOR's Python will have to resort to something like in my suggestion.
2. **For each entry in the dictionary, add another entry using `<component>/<variable>[/prime...]` as the key**, i.e. for each model variable we will have two entries, one with the full path and another with `<component>/<variable>[/prime...]`. This would help @jonrkarr, but I am not sure I like the idea of having two "identical" entries in the dictionary.
3. **Replace each entry in the dictionary with one that uses `<component>/<variable>[/prime...]` as the key**. This might (?) break some existing OpenCOR-based Python scripts, but... those Python scripts will anyway break the day we use libOpenCOR, so... we might as well get things right now.

My preference is for option 3. What about you guys? (FWIW, to implement option 3 would only take a few minutes.)

**Comment by agarny** [comment]
FWIW, I have just implemented option 3.

**Comment by jonrkarr** [comment]
This sounds great. I think this will address the problem I raised.

@bilalshaikh42, this change should fix many of the issues you're seeing with Physiome.

**Comment by agarny** [comment]
> This sounds great. I think this will address the problem I raised.

Excellent. Just waiting for @nickerso and @dbrnz to confirm that they are happy with my choice. Then, I need and want to update various third-party libraries. Once that is done, I will generate a new snapshot.

**Comment by nickerso** [comment]
I guess I am more concerned with not breaking all the current OpenCOR python scripts that people have (including in Physiome articles). The move to libOpenCOR would be a new piece of software right, which people would need to switch to and so know that they need to update things.

A more general solution might be add a new method that will map XPath expressions to OpenCOR keys and then we can move away from everyone needing to do their own ad-hoc method for that. But relies on OpenCOR having access to the original XML file (which I can't remember if that is currently done?).

I'd also be concerned about adding "fake" variables to the `voi_and_variables` list in order to access derivative values. That seems like it should be a different thing.

Having said that, I've had to update my own Python code which uses OpenCOR Python many times over the years, so one more well defined update is not the end of the world.

**Comment by agarny** [comment]
> A more general solution might be add a new method that will map XPath expressions to OpenCOR keys and then we can move away from everyone needing to do their own ad-hoc method for that. But relies on OpenCOR having access to the original XML file (which I can't remember if that is currently done?).

Which original XML file are you talking about here?

To have `[XPath] -> [URI]` could be a good idea for libOpenCOR, but not sure we should go that route for OpenCOR (since it would only be a short-term solution before we switch to libOpenCOR). Also, XPath wouldn't be enough, we would also need to know the degree of the derivative, if any. So, we would probably need something like `[XPath, degree=0] -> [URI]`, i.e. make `degree` optional.

> I'd also be concerned about adding "fake" variables to the `voi_and_variables` list in order to access derivative values. That seems like it should be a different thing.

I was just trying to think about different options, but I agree that this one would be the "wrong" way forward.

> Having said that, I've had to update my own Python code which uses OpenCOR Python many times over the years, so one more well defined update is not the end of the world.

Yeah, sorry about that.

Regarding the Physiome articles, we should be able (and probably should have) a way to check that the scripts still work with the latest version of OpenCOR (or libOpenCOR in the future), and have fix them if ever needed.

Ok, I am going to wait for @dbrnz's view before merging PR #2618.

**Comment by jonrkarr** [comment]
> I guess I am more concerned with not breaking all the current OpenCOR python scripts that people have (including in Physiome articles). 

This issue is motivated by the fact that OpenCOR-python doesn't return some of the data generators involved in a few of the SED-ML files produced by OpenCOR.

@bilalshaikh42 is going to make sure that the revised version of OpenCOR-python is able to simulate each SED-ML file in the latest published changeset of each Physiome record.

**Comment by agarny** [comment]
Thanks in advance @jonrkarr (and @bilalshaikh42), that would be very nice! 👍

**Comment by bilalshaikh42** [comment]
@agarny would it be possible to get a working python release to begin testing and validating the physiome repositories soon?

**Comment by agarny** [comment]
> @agarny would it be possible to get a working python release to begin testing and validating the physiome repositories soon?

Hi @bilalshaikh42, I am working on it. I need to update various third-party libraries. So, everything going well, there will be a new version of OpenCOR sometime this week. Will keep you in touch.

**Comment by bilalshaikh42** [comment]
Sounds good. Wasn't sure if there was a way to build directly from this branch, or would just need to wait for the full OpenCOR release. 
Thanks!

**Comment by agarny** [comment]
At this point, OpenCOR's Python is shipped as part of OpenCOR, so you would need to build all of OpenCOR to get access to its Python (this is a long story as why we originally did it that way, but we are now working on [libOpenCOR](https://github.com/opencor/libopencor) to avoid this problem). It's not difficult (see https://opencor.ws/developer), but probably not worth your time.

**Comment by agarny** [comment]
Sorry @bilalshaikh42, I didn't manage to release a new version of OpenCOR. Still struggling with some third-party libraries that we use... 😕 We have a long Easter weekend here. So, will hopefully have something for you late next week. 🤞

**Comment by bilalshaikh42** [comment]
No worries!

 We have been able to figure out a workaround for now, so this is not a blocker. Since the semantics of the model don't change with the groups, we are just removing the groups for now and simulating the models that way. Once this is updated, we can make sure that we don't lose the group information when importing to BioSimulations

Enjoy the holiday!

**Comment by agarny** [comment]
Closing this issue as a result of PR #2618 having been merged in.

(@bilalshaikh42, I still need to update some third-party libraries. I am currently hoping to generate a new OpenCOR snapshot sometime on Monday 25 April. Sorry for the delay!)

**Comment by agarny** [comment]
@jonrkarr and @bilalshaikh42: I have finally generated a new snapshot for OpenCOR that, among other things, address this particular issue. Sorry for the long delay! Otherwise, as you will have noticed, it now "breaks" your workflow, which I imagine is "normal".
